### PR TITLE
fix(common): prevent injection during jinja templating

### PR DESCRIPTION
**Pull request by davinov**
Forward port of #1849

### CODE REVIEW

File: toucan_connectors/common.py [hunk 1]
```diff
@@ -19,6 +20,11 @@
     import sqlalchemy as sa
 
 
+class NativeImmutableSandboxedEnvironment(NativeEnvironment, ImmutableSandboxedEnvironment): ...
+
+
+jinja_env = ImmutableSandboxedEnvironment()
+
```
Comment (lukapeschke) [inline toucan_connectors/common.py:27]:
```suggestion
```

File: toucan_connectors/common.py [hunk 2]
```diff
@@ -165,9 +171,9 @@ def _render_query(query: dict | list[dict] | tuple | str, parameters: dict | Non
 
         if is_jinja_alone(query):
             clean_p = _prepare_parameters(clean_p)  # type:ignore[assignment]
-            env: Environment | NativeEnvironment = NativeEnvironment()
+            env: Environment | NativeEnvironment = NativeImmutableSandboxedEnvironment()
         else:
-            env = Environment()  # noqa: S701
+            env = jinja_env  # noqa: S701
```
Comment (lukapeschke) [inline toucan_connectors/common.py:176]:
```suggestion
            env = ImmutableSandboxedEnvironment()
```

File: toucan_connectors/common.py [hunk 3]
```diff
@@ -237,7 +243,7 @@ def _flatten_dict(p, parent_key=""):
         parameters.update(p_keep_type)
 
     logging.getLogger(__name__).debug(f"Render query: {query} with parameters {parameters}")
-    return Template(query).render(parameters)
+    return jinja_env.from_string(query).render(parameters)
```
Comment (lukapeschke) [inline toucan_connectors/common.py:246]: Same here, I'm in favor of avoiding globals when possible

File: toucan_connectors/common.py [hunk 4]
```diff
@@ -476,7 +482,7 @@ def pandas_read_sql(
     if convert_to_printf:
         query = convert_to_printf_templating_style(query)
     if render_user:
-        query = Template(query).render({"user": params.get("user", {})})
+        query = jinja_env.from_string(query).render({"user": params.get("user", {})})
```
Comment (lukapeschke) [inline toucan_connectors/common.py:485]: same

**Review by lukapeschke — CHANGES_REQUESTED** [review]
Thanks! Would be great if we could avoid the global var

**Comment by davinov** [comment]
> Thanks! Would be great if we could avoid the global var

Indeed, instantiating an env is cheap, so no need for a global:
[d36ce8f](https://github.com/ToucanToco/toucan-connectors/pull/1850/commits/d36ce8f05d5a9940a14a35b87f6bab89f7e9f878)

**Review by lukapeschke — APPROVED** [review]
Thanks!

**Pull request merged**